### ISSUE TITLE
Intel deflater + long reads data = intermittent corrupt bams

### ISSUE THREAD

**Issue by droazen** [Intel-HLS/GKL#99]
@kvg reports that running the Intel deflater via GATK on long reads data intermittently produces corrupt bam outputs. His specific use case is sharding a single unaligned bam file into multiple smaller bams. Running with the JDK deflater (--use-jdk-deflater) appears to resolve the issue.

Example error when trying to read a corrupt shard (reading with htsjdk produces the same error):

```
$ java -jar gatk.jar SplitSubreadsByZmw -I sharding_test.bam -O intel_compression/v1/ -nr 100000
$ java -jar gatk.jar SplitSubreadsByZmw -I sharding_test.bam -O intel_compression/v2/ -nr 100000
$ java -jar gatk.jar SplitSubreadsByZmw -I sharding_test.bam -O intel_compression/v3/ -nr 100000
$ java -jar gatk.jar SplitSubreadsByZmw -I sharding_test.bam -O intel_compression/v4/ -nr 100000
$ java -jar gatk.jar SplitSubreadsByZmw -I sharding_test.bam -O intel_compression/v5/ -nr 100000

$ samtools view intel_compression/v1/sharding_test.000002.bam > /dev/null
$ samtools view intel_compression/v2/sharding_test.000002.bam > /dev/null
[E::bgzf_read] Read block operation failed with error 2 after 30675 of 72043 bytes
[main_samview] truncated file.
$ samtools view intel_compression/v3/sharding_test.000002.bam > /dev/null
$ samtools view intel_compression/v4/sharding_test.000002.bam > /dev/null
$ samtools view intel_compression/v5/sharding_test.000002.bam > /dev/null

(Only the second attempt yields a corrupted file; runs before and after appear to be correct, despite nothing changing between steps.)
```

There may be a bug in https://github.com/Intel-HLS/GKL/blob/master/src/main/native/compression/IntelDeflater.cc, perhaps triggered when a read spans many compressed blocks.

This bug is also tracked in the GATK repo: https://github.com/broadinstitute/gatk/issues/5798

**Comment by mepowers** [comment]
Hi @droazen, @kvg - Is this is a blocker?  Or is running with the JDK deflater a workaround?

Please feel free to tag me in future GKL issues so I'm sure to see them.

**Comment by kvg** [comment]
@mepowers Running the JDK deflater does permit us to work around the issue for now.  However, the JDK deflater is ~6x slower, so it is unfortunately a rather expensive workaround.

**Comment by mepowers** [comment]
@kvg - thanks for the update.  @droazen mentioned you have a bunch of test cases.  Have you confirmed it's triggered when a read spans many compressed blocks?

**Comment by SnehalAdsule** [comment]
@kvg I am trying to replicate this issue, is SplitSubreadsByZmw a custom option? Is there any other particular branch or fork available for it?

**Comment by SnehalA** [comment]
@kvg I am trying to replicate this issue, is `SplitSubreadsByZmw` a custom option? Is there any other particular branch or fork available with test bam file?

**Comment by kvg** [comment]
Yes - I’ll make the branch and an example available. I have a few deadlines over the next 7 days. I should be able to make both code and data available by Fri Dec 13.

**Comment by SnehalA** [comment]
> Yes - I’ll make the branch and an example available. I have a few deadlines over the next 7 days. I should be able to make both code and data available by Fri Dec 13.
> […](#)
> On Dec 2, 2019, 19:22 -0500, Snehal Adsule ***@***.***>, wrote: @kvg I am trying to replicate this issue, is SplitSubreadsByZmw a custom option? Is there any other particular branch or fork available with test bam file? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.
@kvg IF you has chance to look into this, can you share the branch details. Thanks!

**Comment by kvg** [comment]
Thanks for reminding me - I shall do this tonight.

Thanks,
-Kiran

**Comment by kvg** [comment]
Hi @SnehalA,
  Frustratingly (or thankfully, depending on your point of view), I can't reproduce this error anymore.  I'm not sure *how* it would have been resolved, but it seems to be gone for now.

  FYI, to test this intermittent behavior in the first place, I put together an updated test last night, just running my read sharding code (now called `ShardLongReads`) on a test _E. coli_ sample in loop, testing each iteration to see if any of the output BAMs are corrupted.  The input BAM file and testing script are at: https://www.dropbox.com/sh/9dittkiojm20l9n/AAAda8k7ghOcJzppZ9tWUVt1a?dl=0 .

  The "out.old/" directory (made with GATK SplitSubreadsByZmw version 4.1.0.0-52-g7cc8020-SNAPSHOT) has the results from an older run, the ".000002.bam" being corrupt.  I ran a slightly newer version the sharding tool in a loop 200 times (GATK ShardLongReads version 4.1.0.0-53-g240f7a9-SNAPSHOT).  The newer version doesn't trigger the error.

  I have no idea as to what changed.  My tool is not substantially different between versions: ([old version](https://raw.githubusercontent.com/broadinstitute/gatk/f70eda471e0702b3239eb3ca18352a7d15b0d653/src/main/java/org/broadinstitute/hellbender/tools/walkers/pacbio/SplitSubreadsByZmw.java [new version](https://raw.githubusercontent.com/broadinstitute/gatk/kvg_pbeap/src/main/java/org/broadinstitute/hellbender/tools/walkers/longreads/ShardLongReads.java)

  So, I think I'll update my pipelines to start using the Intel codecs again, and if that more extensivie processing fails for some reason, I'll let you know.  Until then, I think we can close this issue.
